Add Full Support for BYAF Archives

docfail opened the pull request:
# Overview
## What's in this PR
Full support for importing BYAF files without loss of important user data, in accordance with the purpose and specification of the BYAF format. What follows is a list of the supported and unsupported features of BYAF with some explanation regarding which fields are or are not imported. This builds on the work done in PR #4300 but with important changes, the purposes of which will be detailed in various places below.
### BYAF Character Manifest 
<img width="702" height="422" alt="byaf-character-checklist" src="https://github.com/user-attachments/assets/6a847292-6e75-402e-af10-a225ad287a23" />

- The `schemaVersion`, `id`, and `$schema` properties are not imported in any way. These properties have meaning only in the context of the BYAF file itself, and are of no use outside of code intended to parse it, therefore they have no value to a user and may be safely ignored.
- The `name`, `displayName`, `isNSFW`, `persona`, `loreItems`, and `images` are all for the most part preserved (display name is sanitized, and isNSFW is converted to a tag) and imported correctly.
- The timestamps are not preserved, as they would not accurately reflect the date at which the character card in ST was generated.
### BYAF Root Manifest:
<img width="866" height="334" alt="byaf-manifest-checklist" src="https://github.com/user-attachments/assets/471f84c3-c222-46f4-a7be-3f197ed029d9" />

- The `schemaVersion`, `createdAt`, and `$schema` properties are not preserved, as they are specific to the BYAF manifest and are not user data, therefore they are not pertinent to the user performing the import and can be safely ignored.
- The `author` object is full preserved, including the authors name and the URL to their profile where additional cards/BYAF archives may be located. 
- The `characters` and `scenarios` properties are preserved by virtue of importing the character data and all of the scenario data, respectively.
### BYAF Scenario Manifest:
<img width="792" height="699" alt="byaf-scenario-checklist" src="https://github.com/user-attachments/assets/b58401c8-ead0-4507-8593-7046a9d7efab" />

- The `schemaVersion` and `$schema` properties are not preserved as they only have meaning in the context of code performing import/export operations on the BYAF archive, and therefore have no value to the user.
- All remaining properties in the Scenario are preserved and imported correctly.
  - The chat messages are all properly imported, to include scenario, example message, and system prompt overrides per-chat, as well as all swipes information.
  - The settings for the model are preserved in metadata, even though they are not used at this time.
  - The alternative greetings, which represent the possible `firstMessages` from all scenarios are properly imported.
  - The `title` of the chat - while modified to include import timestamps - is preserved and used to title the chats.
  - The background images associated with each chat are imported and linked to the chat via the metadata, even though display of the referenced background image is not implemented at this time.

With these changes, users should be able to export characters with chats into BYAF archives and import them into ST, then continue chatting with characters in ST as though nothing had changed, other than the specific settings for individual AI models on a chat-by-chat basis. 

## What's NOT in this PR
There are feature differences between the Backyard AI Desktop Application and SillyTavern which make a full feature-complete usage of all of the data included impossible at this time. Largely, these features have not been implemented. These include, but may not be limited to:
- Freely changing character icons between multiple variants in a set list saved in the character itself.
- Automatically changing backgrounds when switching chats.
- Changing the model, and all settings for any model automatically when switching chats.
- Blurring NSFW character's icons in the character management panel.
- Displaying a separate display name (such as a full name) from the name the character responds to and uses in chat when viewing characters in the character management panel.
- Choosing whether or not example messages may be pruned from context when the chat data exceeds the context size on a per-chat basis.
- Implementation of example message and system prompt overrides, even though they are saved into the metadata.

## Rationale
The core focus of this PR is to support BYAF files fully in accordance with their intended purpose: Allowing users to migrate character and chat data safely from the Backyard AI desktop application into a different ecosystem. In order to do that, it is critical that **all user data which is needed to continue usage of the characters and chats** is preserved and integrated into SillyTavern's internal formats. Hence, my aim is doing this adequately enough as to be genuinely useful for users who have BYAF files which contain valuable data that they would like to import into SillyTavern, while not attempting to turn SillyTavern into the Backyard AI application. All included (and not included) functionality is done so with this in mind.

## Detailed Summary of changes:
### script.js
 - Added support for both example message and system prompt overrides from chat metadata.
 - Name of user's current persona is now passed to the character import endpoint.
### characters.js
 - To preserve user data from BYAF, character display name (the character's full name usually) is now used to name the character file itself if available. This does not bring feature parity with Backyard AI's character display handling - as ST uses the name from the character card internal data directly in it's sidebar - nor is it intended to, it only preserves the data so that users do not lose it during migration.
 - Additionally, the sanitization now uses a replacement method which utilizes visually-similar characters which are valid on all filesystems supported by the sanitize-filename node library, instead of simply deleting those invalid characters. This is done for better user data-parity.
 - Chats are now imported from the BYAF archive instead of being discarded during character import. The chats also import their overrides settings(all of them) in the chat metadata. The scenario, example message, and system prompt overrides are properly used, however other settings that are intended to change the model or the settings on the model are not used, as this felt out of scope. With the metadata saved, **if** it is desired to support per-chat model settings overrides, then users who have previously imported BYAF characters will see the functionality come online without needing to re-import, meaning they can safely migrate and continue to use the characters in ST without fear of data loss. No editor interface is provided for users to change these overrides from the UI, that may be future work, as I'm not sure how best to display such an editor. They can still change the metadata in the chat files if they wish.
 - Chat backgrounds are now imported from the BYAF archive, instead of being discarded during character import. The chats that used those backgrounds link to (but do not yet **use**) the background in their metadata. Using these backgrounds would be a design change for ST and is squarely out of scope.
 - Character alternative icons are now imported from the BYAF archive instead of being discarded on character import. They are not used however, as doing so requires changing how the characters are saved entirely, potentially to use the new CharX format, which felt out of scope. 
### byaf.js
- The replaceMacros() and formatExampleMessages() methods are now static, as they are needed within the static context of chat file generation, and don't actually use any non-static fields. Usage has been updated to reflect this.
- The formatAlternateGreetings() now takes an array of all scenarios in the BYAF, and returns all possible greetings from the archive. The prior implementation of the method would always return an empty array for a valid BYAF file, because it trimmed the first example of the first scenario and used "the rest", but according to the spec, [any individual scenario's "firstMessages" array "MUST contain **at most one** message"](https://github.com/ahoylabs/byaf/blob/main/specs/v1/scenario.md#first-messages). Backyard AI does not support a cycling list of first messages/greetings, instead every chat has the potential to have a unique greeting or not. The BYAF file's format reflects this. As such, to get the actual list of "all possible greetings", it is necessary to pull the greeting from all existing scenarios.
- The getScenarioFromManifest() is now getScenariosFromManifest() to reflect it's new function.
- The getScenariosFromManifest() now retrieves *all* BYAF Scenarios from the archive, rather than discarding all but the first one. A "scenario" in the context of a BYAF archive does NOT mean the same thing as it does in the context of ST. A scenario in a BYAF archive is a user's chat history for a specific chat with a specific character, including the "narrative"(which is what ST, and Backyard AI both refer to as the "scenario"), the first message, the example messages, the system prompt, and the model and generation settings for that chat. This is crucial data for users with BYAF exported characters and should not be discarded.
- The getCharacterImage() is now getCharacterImages() to reflect it's new function.
- The getCharacterImages() now retrieves all character icon images from the BYAF archive instead of grabbing the first and discarding the rest. This is needed so they can be saved to file during the character import process. 
- The getCharacterCard() method now takes an array of all scenarios instead of just one. It does this to use the new form of formatAlternateGreetings() to get all of the alternative greetings for the character.
- The getCharacterCard() method no longer sanitizes the characters name before saving it into the card itself, as the sanitization is only relevant for filenames. The sanitization is now performed when generating the filename for the character instead, which is subsequently used for all file-related matters.
- The getCharacterCard() method now saves the author's url in the author comments field, since that wasn't actually being used and there's no author comments to save from the BYAF. This makes it a convenient place to prevent this data from being lost, and thereby attributes characters appropriately to their creators on import.
- The getCharacterCard() method now saves the isNSFW flag by adding a "nsfw" tag to the character card. Since BYAF doesn't have tags, the tags were previously unused, making this a convenient place to preserve this character data.
- The new getChatBackgrounds() method returns all of the chat backgrounds saved in the BYAF archive, mapping them to their internal reference name so that the chat metadata can be written to reflect their new storage locations once imported. This means backgrounds no longer need to be discarded during BYAF import.
- The new static getChatFromScenario() method allows retrieval of chats from a scenario file, formatted for ST, so that chats no longer need to be discarded during BYAF import. This method imports all chat messages, adds the appropriate first message, sets the scenario, example message, and system prompt overrides, as well as importing all other ancillary metadata from the chat which ST may or may not use in the future. When the chats are built, swipes are all included and the correct swipe from the original conversation will be selected for all chat messages, including the most recent message from the AI. This allows users to take full advantage of the branching feature in ST that is not supported in the Backyard AI desktop application. This has been tested to be in full parity with the chat state as it was in the Backyard AI application prior to import. This method also includes functionality to interleave chat messages, repairing bad chat data from BYAF imports where it is heuristically appropriate to do so. The method is static to prevent the parser object from needing to perform the actual duties of creating chat files on the user's system and the like, which is out of the scope of the responsibilities of a parser.
- The parse() method now retrieves all of the chat backgrounds and returns them before exiting.
- The parse() method's signature has been updated to reflect the all of the new functionality.
### util.js
- Migrated a copy of the getUniqueName() utility function from the frontend version of util.js
### byaf.d.ts
- Created ByafChatBackground type to simplify reading code surrounding chat background logic.
- Created ByafParseResult to condense the now unwieldy anonymous return type from ByafParser.parse()

## Testing this PR:
[Noor-Shepard-Edwin-Evy-Test-Characters.zip](https://github.com/user-attachments/files/23154025/Noor-Shepard-Edwin-Evy-Test-Characters.zip)

Use the provided BYAF files, perform an import and verify the following:
- Within the user character directory, the alt icons for each character that have them have been saved.
- All chats have been imported from the archive, and the metadata for the generated chats includes the appropriate overrides and displays the correct chat background. 


<!-- Put X in the box below to confirm -->

## Checklist:

- [x] I have read the [Contribution guidelines](https://github.com/SillyTavern/SillyTavern/blob/release/CONTRIBUTING.md).

docfail committed:
+ Fix discarding all scenarios aside from the first one, which would cause all chats other than the first one to be lost.
+ Accumulate alternative starting messages from all scenarios, as different scenarios are how Backyard handles having alternative greetings. A scenario with multiple messages in its "firstMessages" array is invalid as per BYAF spec.
+ Return scenarios so that they can be used to populate chats, scenario overrides, and in the future, the other character settings overrides as well.

docfail committed:
+ Added support for importing chats (including swipes) from BYAF. This brings the core functionality of BYAF imports mostly in-line with the standard, without the inclusion of any features that are currently missing from ST.
> The only supported feature not yet implemented is scenario overrides per-chat. This will be the the goal of the next commit.

docfail committed:
+ Implemented scenario overrides, fully bringing ST's BYAF import as in-line with the BYAF standard as possible without adding new feature functionality to ST.
+ Also added some new chat metadata fields in preparation for example message and system prompt overrides.

docfail committed:
+ Added support for pulling alternate icons from the BYAF archive. They are not used yet, but they are safely stored.
+ Added support for pulling backgrounds from the BYAF archive. They are not used yet but they are stored and linked to the chats they originally belonged to.
+ Saved every remaining field that was previously discarded from BYAF that a user actually needs in a way that could be used in the future, and can be accessed by the user if needed after import. Any fields that are not preserved at this point are specfic to either Backyard AI's desktop application, or to the BYAF manifest itself, and therefore have no relevance to a user.
> With this, BYAF import should be fully supported and in-line with the official spec.

docfail committed:
+ Updated comments and method names to reflect changes appropriately. First round of touch ups.

docfail committed:
+ More comment updates.

docfail committed:
- Removed commented code, that is no longer needed for reference.

docfail committed:
> No longer sanitize for filenames in the character card, as there's no need. The sanitization now happens when generating the filename, which is subsequently used for file operations.
+ Added replacement function for sanitizing filenames which uses visually similar but valid for filenames replacement characters instead of deleting the characters for better data parity.

docfail committed:
+ Preserve isNSFW flag as a tag, so it's not lost on import.

docfail committed:
+ preserve labels on character alt avatars

docfail committed:
+ Deduplicate first messages, so that 5 chats with the same first message don't create 5 identical first messages.

docfail committed:
+ Adjusted icon naming logic to produce more sensible results.

Cohee1207 commented on public/script.js:
```diff
@@ -2690,11 +2690,14 @@ export function getCharacterCardFields({ chid = null } = {}) {
     }
 
     const scenarioText = chat_metadata['scenario'] || character.scenario || '';
+    const exampleDialog = chat_metadata['mes_example'] || character.mes_example || '';
```
Keep in mind that in group chats this will cause ALL characters to have the same example dialog.

docfail replied:
Can you provide an example of what a group chat file looks like?

Cohee1207 replied:
Characters in group chats share the chat history but either swap character cards (default) on every prompt or join them. Group chat files can also have metadata, which is shared between all characters in prompts.

https://docs.sillytavern.app/usage/core-concepts/groupchats/

docfail replied:
Marking this as resolved for now as this code has been moved to a new branch for the not-yet-made character settings overrides PR. I'll take the feedback into account for that PR, though I might ask for suggestions if nothing neat comes to mind.

Cohee1207 commented on src/byaf.js:
```diff
@@ -274,18 +320,122 @@ export class ByafParser {
+                mes_example: ByafParser.formatExampleMessages(scenario?.exampleMessages),
+                system_prompt: ByafParser.replaceMacros(scenario?.formattingInstructions),
+                mes_examples_optional: scenario?.canDeleteExampleMessages || false,
+                model_settings: {
```
This should not be imported. Coupling characters with model sampling settings is not a good thing even in concept, I do not agree that this is needed. Not all models and APIs support the same settings, and not all of them behave similarly even if the settings are the same.

docfail replied:
I'm afraid I have to disagree entirely. 
1. The model settings are not implied to use the same settings, the model itself is included. So models and APIs not using the same settings doesn't make a difference. At the moment, this data isn't used at all anyway.
2. I personally am not attached to this particular user data, and originally I wasn't going to include it but it's in the specification because it was important data for the chats. Some users have spent a lot of time meticulously tweaking the models and settings for their characters and chats. To them, this would be tantamount to ST throwing out potentially hundreds of hours of work. It may not be a clean transition without them functioning, but if they are imported, then at least the data cannot be said to be lost.

If ST is going to support BYAF files, it should support them in a way that aligns with their intended purpose, which is safely saving the character and chat data such that users can migrate into a new ecosystem. As I outlined above in the PR, the BYAF format has a purpose, and user data should not be thrown out without their input.

Cohee1207 replied:
My reasoning is that ST is not a drop-in replacement for Backyard, so that 1:1 matching between features is not expected. I don't see much value is saving data that isn't going to be used, but if you need to preserve it, make it clear that it's from BYAF and not a part of data that is going to be used natively, i.e. by adding a `byaf_` prefix: `byaf_model_settings`.

docfail replied:
I'm very comfortable with that compromise. Updating now, I'll resolve when I've pushed it.

docfail replied:
Updated.

Cohee1207 commented on src/endpoints/characters.js:
```diff
@@ -806,10 +806,85 @@ async function importFromByaf(uploadPath, { request }, preservedFileName) {
     await fsPromises.unlink(uploadPath);
     console.info('Importing from BYAF');
 
+    const custom_sanitize_safe_replacement = (char) => {
```
What's the significance of this? The default sanitizer should work fine and not create any confusion in naming.

docfail replied:
The default sanitizer mangles a lot of names for characters I personally have and have seen, this is because it completely deletes characters that are not valid, instead of providing replacements. Backyard did not have a requirement that character's names be able to be used as a filename, so users often used character names that did not align with file naming requirements.

Cohee1207 replied:
This behavior is fine and expected. If you'd like a change, you either need to implement it everywhere the santizer is used or drop it from here. Making a change only in one place is not great for consistency. This a good candidate for a separate discussion and its own PR.

Cohee1207 replied:
I see the comment resolved, but what is the decision here regarding the sanitation rules?

docfail replied:
I would actually prefer to change it everywhere to be honest. Only issue is, that would be changing it in 83 places across 20 files(the sanitizer is imported from an external package, so I cannot modify it's source). Way too much bloat to add to this PR. Would making a defined function in the `utils.js` out of the replacement function and using that here be a reasonable compromise? Then a follow-up PR can change the usage across the entire project to use that defined replacement function? I'd prefer not to leave the single lingering issue of mangled character full names in if possible.

docfail replied:
Update: I've done just this, take a look.

Cohee1207 replied:
After giving it some thought, there are reasons why I wouldn't want to change it everywhere. This has an unknown and potentially drastic level of impact. Some areas of functionality depend on having the sanitation rules stable, for example: access to character sprites is performed using a sanitized character name (this also means that import of character icons needs to use card name instead of file name, I'll add a comment there). Changing the sanitation rules means that names containing special characters will stop resolving to the same paths they used to resolve, which will break file references. I don't think that using it to define a file name on import will break anything, but I'd prefer a more standard way of replacing invalid characters, such as a standard ASCII underline (`_`) instead of rare Unicode characters.

docfail replied:
Taking a look at that sprites code, I'm not sure I really like the logic there if I am honest. It seems unnecessarily risky imo to perform arbitrary modifications(partially dependent upon code ST doesn't control) on available data to *locate* the folder something is stored in *every time*. It would be better to generate the sprites' path once, store that somewhere, and reuse it *without* modification every time it's requested, both to be more efficient, and to prevent issues like this, or issues where two characters names' which differ only by an invalid character (e.g. "Andrew" and "Andrew?") would resolve to the same sprite path.

That being said, that's neither here nor there for this discussion. Changing that logic would be out of scope for this PR, and almost certainly warrants a detailed discussion on exactly how it would be changed and what would be done about existing data. I just wanted to voice my thoughts on the matter after looking at it.

docfail replied:
Back on topic though: I don't think the benefits of using "_" outweigh the benefits of using the common lookalikes.
Using common lookalikes:
- The full name of the character is visually preserved as well as it can be.
- The name is perfectly safe on any standard filesystem where sanitize-filename is expected to sanitize for
- There are less collisions between characters whose names may differ only by illegal characters
- In the absolute worst case, where the font on the system doesn't have an entry for the character, a placeholder will be displayed, which has the same effect as "_" but without increasing collisions and mangling names more for systems that can display the characters.

Using just "\_":
- A simple ASCII character is used instead of a utf-8 character for these characters, but this doesn't affect all of the existing utf-8 characters.
- The name is also perfectly safe on any standard filesystem where sanitize-filename is expected to sanitize for

docfail replied:
Update: I have preserved the display name property in the extensions of the card so it matters less now how well it's preserved here. If you're adamant about using "_", I can live with that, just let me know.

Cohee1207 replied:
Ok, but I'll replace the property `full_name` with `display_name`. Not sure how and when this will be implemented (I had some code stashed long ago regarding display names due to being unclear where exactly to apply it), but `display_name` looks better to me.

Cohee1207 replied:
I picked _ because that's how I know browsers are handling invalid characters in names. Easy to test, paste in DevTools and run:
```
const blob = new Blob([]);
const a = document.createElement('a');
a.download = '<>:"|*?.bin';
a.href = URL.createObjectURL(blob);
a.click();
```

docfail replied:
I was debating between `display_name` and `full_name`. tbh I was worried display names wouldn't ever get implemented so I went with `full_name` to avoid confusion but if you're considering implementing display names, `display_name` is my preference anyway. As an aside: I'd be happy to discuss display names in a discussion for a future PR, it would be a very nice feature to have IMO.

docfail replied:
Regarding the way \_ is used by browsers handling invalid characters, that makes sense. I did test it and see that it created "______.bin". I'm good with changing to "\_" for the replacements to be consistent with that.

docfail replied:
Updated. Now it uses "\_" for all replacements.

Cohee1207 replied:
The whole display name debacle is around where exactly to apply it. We can't just do as "they" do in certain other frontends just because ST has much-much more features to worry about and lots of them operate with a character name. We could limit it to just and exclusively the character list, leaving the in-chat name, {{char}} macro, etc. intact.

docfail replied:
https://github.com/SillyTavern/SillyTavern/discussions/4715
Here, I started a discussion so we can go a bit further with it and I can resolve this thread.

Cohee1207 commented on src/endpoints/characters.js:
```diff
@@ -806,10 +806,85 @@ async function importFromByaf(uploadPath, { request }, preservedFileName) {
+        // This will not yet allow changing icons within the UI but at least the icons will be available for manual selection, rather than being lost.
+        const altImagesFolder = path.join(request.user.directories.characters, fileName.replace('.png', ''));
+        if (!fs.existsSync(altImagesFolder)) fs.mkdirSync(altImagesFolder, { recursive: true });
+        const extension = icon.filename.split('.').pop() || 'png';
```
Please use `path.extname` and `path.basename` for these operations.

docfail replied:
I was not aware those existed. Thank you!

docfail replied:
Done, thanks!

Cohee1207 commented on public/script.js:
```diff
@@ -2690,11 +2690,14 @@ export function getCharacterCardFields({ chid = null } = {}) {
 
     const scenarioText = chat_metadata['scenario'] || character.scenario || '';
+    const exampleDialog = chat_metadata['mes_example'] || character.mes_example || '';
+    const systemPrompt = chat_metadata['system_prompt'] || character.data?.system_prompt || '';
```
A user would have no ability to know there's a per-chat override of system prompt or even remove it without editing a chat file. This needs a UI counterpart. And it's better to split an ability to override prompts and other data into a separate PR from BYAF import.

docfail replied:
I was intending to add a UI counterpart as a follow up PR, the reason this was included now is so that users don't need to wait for 2 PRs to get through before they are able to make use of their chats.

Cohee1207 replied:
I don't see a problem in such split. staging is a DIY branch, users who need stability and feature completeness will need to wait for a release increment, which takes as long as it needs to get finished.

docfail replied:
Okay. that makes sense. I'll move this change to a different branch and add the UI alongside them, and make a new PR for that. While I have you, do you have a recommendation on what the UI for that should look like? Should there be two new options in the dropdown alongside the scenario override or should these be merged into the scenario override menu, turning the scenario override menu into a "character settings overrides" menu?

Cohee1207 replied:
New fields (if you decide to allow override of example messages too) alongside scenario override are fine. "character settings overrides" is fine for a new dialog name.

docfail replied:
This part of the changes have been moved to a new branch specific for this new character settings override UI feature.

Cohee1207 commented on src/byaf.js:
```diff
@@ -53,29 +52,37 @@ export class ByafParser {
+            // As per the BYAF spec, "firstMessages" array MUST contain AT MOST one message.
+            // So we only consider the first one if it exists.
+            const firstMessage = scenario?.firstMessages?.[0];
+            if (firstMessage?.text && firstMessage.text !== scenarios?.[0].firstMessages?.[0].text && !greetings.includes(ByafParser.replaceMacros(firstMessage.text))) {
```
You can avoid having to check for duplicates by declaring `greetings` as `new Set()` and converting it to array before returning.

docfail replied:
That's good to know, I'll do just that.

docfail replied:
Done. Thank you!

Cohee1207 commented on src/byaf.js:
```diff
@@ -274,18 +320,122 @@ export class ByafParser {
+            });
+        }
+
+        const sort_by_timestamp = (newest, curr) => {
```
Please use camelCase for local function names.

docfail replied:
Done. Thank you.

Cohee1207 commented on src/byaf.js:
```diff
@@ -274,18 +320,122 @@ export class ByafParser {
+            return aiMessage.outputs.map(output => output.text);
+        };
+
+        const fix_prior_bad_backyard_imports = true; // If true, reorders messages by interleaving user and character messages so that they are in correct chronological order. This is only needed to import old chats from Backyard AI that were incorrectly imported by an earlier version that completely messed up the order of messages. Backyard AI Windows frontend never supported creation of chats w ...
```
This is a const, so it will always be true. Is this just to add a comment? Or is it meant to be configurable?

docfail replied:
Primarily to add a comment. I'm open to it being configurable, but I didn't really see the need for making it so at the moment, as I can't see a case where a user would want to actively leave chats mangled that otherwise make more sense in order.

docfail replied:
As this seems to have been just a question, I'm going to mark this as resolved to keep track of updates. If you would like to request a change here please feel free to unresolve and clarify what you would like me to do.

Cohee1207 replied:
I don't see a reason to add a variable for a comment. Just place a block comment above the condition.

docfail replied:
I agree with that, I was fence sitting on the matter of whether it should be configurable or not but I suppose since I don't see many likely scenarios where it was needed, if it becomes an issue a configuration value can easily be added in the future. Updated.

Cohee1207 commented on src/endpoints/characters.js:
```diff
@@ -806,10 +806,85 @@ async function importFromByaf(uploadPath, { request }, preservedFileName) {
+
+    let bgIter = 1;
+    // Upload backgrounds
+    for (const bg of byafData.chatBackgrounds) {
```
This will populate a global list of backgrounds. We have support for per-chat backgrounds, but not per-character backgrounds. You can use `userImages` directory to store custom images.

Cohee1207 replied:
If you need references, look at how `public/scripts/backgrounds.js` interact with the image generation extension for generating chat backgrounds.

docfail replied:
Ah, I was unable to get the image generation extension running to verify that it generated per-chat backgrounds, so I wasn't able to test if it did that or not. That's good to know. I'll change the directory to match that instead, thank you.

docfail replied:
Updated, now it uses the existing system and I can see the chat backgrounds working out of the box. Thank you for letting me know this functionality exists, it's a huge improvement.

Cohee1207 commented on src/endpoints/characters.js:
```diff
@@ -806,10 +806,85 @@ async function importFromByaf(uploadPath, { request }, preservedFileName) {
+
+
+    // Save alternate icons for the character.
+    for (const icon of byafData.images.slice(1)) {
```
Importing chats, images, etc. on every import will create duplicate files if the Replace / Update function is used on a character. This is undesirable. Replace / Update should not add other files except the character card PNG.

docfail replied:
Not importing chats, images, etc is unacceptable, at that point there is very little reason to support BYAF files at all. What would be your suggestion? Is there a way to check if we're doing Update/Replace rather than importing a new character? Does having a preservedFileName set imply that this is an update/replace operation?

docfail replied:
Upon looking further, I don't see a "Replace/Update" function anywhere in the UI. Where is that?

Cohee1207 replied:
> Does having a preservedFileName set imply that this is an update/replace operation?

Yes. My proposal is not to import supplementary files if it's set.

> Upon looking further, I don't see a "Replace/Update" function anywhere in the UI. Where is that?

"More..." dropdown menu in the character right panel.

docfail replied:
Okay, I couldn't find it in the UI but I do see a log that seems to imply - following the logic - that "preservedFileName" being set does in fact imply that the character is being "replaced or updated" so I'll go with that assumption and update the byaf character import logic accordingly. then test to verify it doesn't cause problems for normal imports and if not I'll push that out.
Edit: I wrote this before seeing your latest comment, thank you for letting me know where to find that.

docfail replied:
Okay, I've updated the logic so that if preservedFileName is set, the supplementary files are not imported. Thank you for letting me know.

Cohee1207 commented on src/endpoints/characters.js:
```diff
@@ -806,10 +806,85 @@ async function importFromByaf(uploadPath, { request }, preservedFileName) {
+        const extension = bg.prev_paths?.[0]?.split('.').pop() || 'png';
+        const baseName = `${fileName.replace('.png', '')}_bg_`;
+        let file = baseName + bgIter;
+        while (fs.existsSync(path.join(request.user.directories.backgrounds, `${file}.${extension}`))) {
```
Duplicated logic of getting unique file names should be refactored into a utility function, I think a similar one already exists even.

docfail replied:
If one exists, I'd love you to point me to it before I duplicate it unintentionally.

Cohee1207 replied:
I found it in the frontend utils, so it should be copied into backend utils (reusing code between them is not possible)

https://github.com/SillyTavern/SillyTavern/blob/e9cd32a123f49c1f5d2f42f4320021007c1e9a50/public/scripts/utils.js#L638-L646

docfail replied:
Convenient, thanks. I'll copy this into the backend utils.js and refactor the code to take advantage.

docfail replied:
Okay, done.

Cohee1207 commented on src/endpoints/characters.js:
```diff
@@ -806,10 +806,85 @@ async function importFromByaf(uploadPath, { request }, preservedFileName) {
+    /**
+     * @param {Partial<ByafScenario>} scenario
+    */
+    const createChatAsCurrentPersona = (scenario) => {
```
Persona in ST refers to **user** persona only. This creates chat for a character, not persona.

docfail replied:
The chat files include the user's persona as the name. The function is "create chat as current persona". It's implied that it's creating a chat for this character, but ST's chat format doesn't use placeholders, it uses the actual name of the persona in the files, hence, this is creating the chat *as* the current user persona. I called it out because otherwise it might have been unclear why it was necessary to append the user name to the body of the request.

Cohee1207 replied:
To be honest, nothing breaks if you omit user_name, character_name and create_date fields for chat files. They are not used anywhere and are legacy fields from a pre-fork age. Only chat_metadata matters. I just never got to removing them.

docfail replied:
What about the "name" entries in all of the messages? Are those also superflous? If so that would simplify this quite a lot. It's been a couple of days, but from my original testing early on I seem to remember that ST does actually use the "name" on user's messages, and will not simply ignore them in the chat window, which is the primary reason this was included, not for the user_name field in the metadata block.

docfail replied:
I have just tested again to re-verify, and not using the user name does in fact still cause problems. Whether or not the initial block needs a value in `user_name` notwithstanding, the UI *does* in fact use the value in the "name" field of every user message. 
As proof:
<img width="932" height="608" alt="available-personas" src="https://github.com/user-attachments/assets/95d59aef-373b-4cf3-83fd-a3d794fea416" />

Here are my available personas. You'll note that "{user}" and "" are not personas I have.

<img width="932" height="572" alt="with-request body user_name" src="https://github.com/user-attachments/assets/15c96629-d5ae-4ee5-88da-a764cb657e87" />

Here is what it looks like when I am using the user name from the request body.

<img width="924" height="558" alt="without-request body user_name" src="https://github.com/user-attachments/assets/c5e978d9-4fa6-4d35-ae1e-64f9a4fea7e4" />

Here is what it looks like when I use {user} - something that seems like it could make a sensible replacement for the persona name.

<img width="930" height="572" alt="without-request body user_name-trying-something-falsy" src="https://github.com/user-attachments/assets/458f5869-3ae6-4d6c-8f40-98088d4e1e51" />

And just to really make sure, I even tried something falsy - namely ''

Cohee1207 replied:
So, to summarize this, `name` field in messages is needed, but both chat header fields are not needed?

docfail replied:
I didn't test if the chat headers aren't needed, as it wouldn't affect the code here whether it were needed or not, but the `name` in the messages is needed, yes.

Cohee1207 reviewed (changes requested):
A solid start, but some improvements are required.

I also recommend splitting into smaller, more focused PRs, such as (just an example, not a call to action):

- Support for per-chat prompt override
- Support for multiple scenarios
- Support for multiple images
- Import of chats
- Import of backgrounds
- etc.

This will make it easier for me to review and test the changes. Also, please provide example files for testing. I don't have any BYAF files to test on.

docfail commented:
> A solid start, but some improvements are required.
> 
> I also recommend splitting into smaller, more focused PRs, such as (just an example, not a call to action):
> 
>     * Support for per-chat prompt override
> 
>     * Support for multiple scenarios
> 
>     * Support for multiple images
> 
>     * Import of chats
> 
>     * Import of backgrounds
> 
>     * etc.
> 
> 
> This will make it easier for me to review and test the changes. Also, please provide example files for testing. I don't have any BYAF files to test on.

I actually did include example files which test all features of the PR but I guess I accidentally deleted that section before I submitted the PR. I have no idea how I did that, one moment while I rectify that. EDIT: done, you're good to go there.

Also, while I understand the desire to cut down the PR, I would prefer not to separate out implementation of the core functionality of BYAF import into multiple separate PR's. BYAF import is one feature, these changes are aligning that single feature with the specification. 
For examples of how this rationale affects potential splits, using those potential suggestions provided:
- I would consider "support for per chat prompt override" and "support for multiple example messages" to be PR's that added a UI to interact with this metadata. 
- Import of chats must necessarily include all of this metadata (backgrounds, scenario overrides, example message overrides, and prompt  and settings overrides all included) because otherwise, users importing chats would lose it if they imported the chats before the other metadata features were merged in.

docfail committed:
> User path.basename and path.extname for getting the base name and extension of files, respectively.

docfail committed:
> Should use camelCase for local function names.

docfail committed:
> Use a set for greetings instead of explicitly checking for duplicates.

docfail committed:
> Missed a snake case local function name.

docfail committed:
> Updated background import logic to integrate with existing per-chat background system.

Cohee1207 commented on src/byaf.js:
```diff
@@ -274,18 +320,124 @@ export class ByafParser {
+                mes_examples_optional: scenario?.canDeleteExampleMessages || false,
+                model_settings: {
+                    model: scenario?.model || '',
+                    temperature: scenario?.temperature || 1.2,
```
This code will set a default value to sampler params such temperature, top k, top p, min p, etc. if their value is 0, which is a totally legit value for sampling params. Use `??` operator instead.

docfail replied:
That makes sense, thank you!

docfail replied:
Updated. Thanks again, I should have caught that myself.

docfail committed:
+ Updated logic to skip import of chats and backgrounds when a character is being updated/replaced.

docfail committed:
+ Added copy of getUniqueName helper function from frontend utils to backend utils.
+ Refactored duplicated logic for getting unique filenames for backgrounds and character alt icons to use the newly available function.

docfail committed:
> fixed adding an extra dot to the filenames for icons after switching to path.extname for getting the extension

docfail committed:
> Updated metadata settings generation in byaf chat so that it doesn't overwrite valid yet 'falsy' values with defaults. Oops

docfail committed:
> Remove variable for configuration that will likely never need to exist.

Cohee1207 commented on src/byaf.js:
```diff
@@ -254,6 +269,37 @@ export class ByafParser {
             create_date: humanizedISO8601DateTime(),
```
When importing chats and opening the character for the first time, it will not open an imported chat, instead it will create a new chat with a timestamp name. However, it's possible to use a non-standard spec extension (`chat`) in the object root to preset a file name that will be used when opening a character for the first time. You just need to defer writing the character data to a file until chat files have been written, and specify a file name (extensionless) in the JSON data.

docfail replied:
is that filename assumed to exist in the chats/character name/ directory? Or will that need to be included in the path for `chat` ?

docfail replied:
Nevermind, I figured it out. Updated.
This is a huge QoL improvement, tbh. I was pretty annoyed that it kept opening a newly created chat when I was testing import instead of one of the imported ones. Thank you for this suggestion.

Cohee1207 commented on src/byaf.js:
```diff
@@ -274,18 +320,129 @@ export class ByafParser {
     /**
      * Parses the BYAF data.
-     * @return {Promise<{card: TavernCardV2, image: Buffer}>} Parsed character card and image buffer
+     * @return {Promise<{card: TavernCardV2, images: {filename: string, image: Buffer, label: string}[], scenarios: Partial<ByafScenario>[], chatBackgrounds: Array<{name:string, data:Buffer, prev_paths:string[]}>, character: ByafCharacter}>} Parsed character card and image buffer
```
This return type is getting crowded. It's better to create a dedicated typedef for all these fields, like `ByafParseResult`

docfail replied:
Thank you for suggesting a name! Updated.

Cohee1207 commented on src/byaf.js:
```diff
@@ -254,6 +269,37 @@ export class ByafParser {
+     * Gets chat backgrounds from BYAF data mapped to their respective scenarios.
+     * @param {ByafCharacter} character Character object
+     * @param {Partial<ByafScenario>[]} scenarios Scenarios array
+     * @returns {Promise<Array<{name:string, data:Buffer, prev_paths:string[]}>>} Chat backgrounds
```
It's better to create a type definition for it similar to ByafCharacter / ByafScenario, e.g. ByafChatBackground. This is better for DRY also.

docfail replied:
Couldn't agree more, updated. Code looks much cleaner now, thanks.

Cohee1207 commented on src/byaf.js:
```diff
@@ -274,18 +320,129 @@ export class ByafParser {
+                    grammar: scenario?.grammar ?? null,
+                },
+                chat_backgrounds: [chat_background],
+                custom_background: `url("${encodeURI(chat_background)}")`,
```
Do BYAF files _always_ contain a background? If chat_background is an invalid file path, it will cause a broken image to display. `custom_background` should be an empty string if `chat_background` is falsy.

docfail replied:
Sounds good to me. Updated.

Cohee1207 commented on src/byaf.js:
```diff
@@ -274,18 +320,129 @@ export class ByafParser {
+                    by_prompt_template: scenario?.promptTemplate ?? 'general',
+                    grammar: scenario?.grammar ?? null,
+                },
+                chat_backgrounds: [chat_background],
```
Make an array empty if `chat_background` is falsy.

docfail replied:
Updated.

Cohee1207 commented on src/byaf.js:
```diff
@@ -274,18 +320,129 @@ export class ByafParser {
+     * @returns {string} Chat data
+     */
+    static getChatFromScenario(scenario, userName, characterName, chatBackgrounds) {
+        const chat_start_date = scenario?.messages?.length == 0 ? humanizedISO8601DateTime() : scenario?.messages?.filter(m => 'createdAt' in m)[0].createdAt;
```
Prefer camelCase for local variable names. Same below.

docfail replied:
Updated.

Cohee1207 commented on src/endpoints/characters.js:
```diff
@@ -804,10 +804,82 @@ async function importFromByaf(uploadPath, { request }, preservedFileName) {
+        if (Buffer.isBuffer(bg.data)) {
+            const newFile = `${file}${extension}`;
+            writeFileAtomicSync(path.join(filePath, newFile), bg.data);
+            bg.name = clientRelativePath(request.user.directories.root, path.join(filePath, newFile)); // Update background name to the new file
```
Is this updated `bg.name` value used anywhere after updating? I couldn't find any clues for that.

docfail replied:
Yep, the bg.name is actually what gets used to create the reference to the background image in the chat metadata. It's pretty sneakily hidden in there on line 333 of byaf.js right near the top of getChatFromScenario.

Cohee1207 commented on src/endpoints/characters.js:
```diff
@@ -804,10 +804,82 @@ async function importFromByaf(uploadPath, { request }, preservedFileName) {
+        const extension = path.extname(bg.prev_paths?.[0]) || '.png';
+        const baseName = `${path.basename(fileName)}_bg`;
+        const filePath = path.join(request.user.directories.userImages, fileName);
+        if (!fs.existsSync(filePath)) fs.mkdirSync(filePath, { recursive: true });
```
This condition will always be false as you're checking for file existence, not directory. Also `userImages` will always exist (it's auto-created on startup) unless you manage to delete it in runtime. Same comment probably applies to other areas.

docfail replied:
The name "filePath" may be a little misleading here. I am actually checking for a directory here, not a file. The images are stored in user/images/<character_name>/ and that's what this check is verifying the existence of, so that writeFileAtomicSync doesn't throw an error when trying to create a temp file.

Cohee1207 commented on src/byaf.js:
```diff
@@ -254,6 +269,37 @@ export class ByafParser {
+                    backgrounds.push({
+                        name: `${character?.name} bg ${i++}` || '',
+                        data: data,
+                        prev_paths: [bgImagePath],
```
I don't get why is it called prev_paths and why is it an array. Can you clarify that?

docfail replied:
The reason is that multiple chats may use the same background image. I did not want to save multiple duplicate copies of the same background image, even though the BYAF archive does so, so I compare the backgrounds when enumerating them and only create one image when multiple chats use the same image byte-for-byte. The array here stores the original references so that when creating the chat metadata, it can point to the correct imported background. "prev paths" is just saying these were previous paths pointing to this background.

docfail replied:
For quicker reference, the code that condenses duplicate backgrounds is at line 288 just above in getChatBackrounds() here, and the prev_path resolving to the new filename is 333 at the top of getChatFromScenario  a little below.

Cohee1207 commented on public/script.js:
```diff
@@ -8893,6 +8893,7 @@ async function importCharacter(file, { preserveFileName = '', importTags = false
     const formData = new FormData();
     formData.append('avatar', file);
     formData.append('file_type', format);
+    formData.append('user_name', name1);
```
This is not really needed. See my other comment regarding user/char names in chat files.

docfail replied:
As far as I can tell from testing, it is still very much needed. See my reply to your comment in the other thread.

docfail commented:
Created #4707 for the overrides updates needed to use the metadata.

docfail committed:
> custom_background should be an empty string if chat_background is falsy

docfail committed:
> Make an array empty if chat_background is falsy.

docfail committed:
> Prefer camelCase for local variable names.

docfail committed:
+ Added definitions for ByafParseResult and ByafChatBackground.
> Updated overused and unweildy anonymous types to the newly defined types.

docfail committed:
+ When importing a character with chats, the character will now no longer open to a new empty chat, but instead will pick the first of the imported chats to open up to.

Cohee1207 commented:
The code looks good, and I can get to testing it soon, I only have one question still unresolved regarding the name sanitation. Let me just copy it here so it wouldn't get buried:

> If you'd like a change, you either need to implement it everywhere the sanitizer is used or drop it from here. Making a change only in one place is not great for consistency. This a good candidate for a separate discussion and its own PR.

Cohee1207 committed:
Merge branch 'staging' into full-byaf-support

Cohee1207 committed:
Rename prev_paths => paths

Cohee1207 commented:
`prev_paths` still seemed wrong to me, so I renamed it to just `paths`. Should be good enough.

Cohee1207 committed:
Prettify logs display

Cohee1207 committed:
Set fallback chat name if scenario is untitled

Cohee1207 committed:
card.data.name => card.name

Cohee1207 committed:
Optimize first scenario message access

Cohee1207 reviewed (approved):
Tested, looks fine. Ready to merge when the sanitizer question is resolved.

docfail commented:
Replied in sanitizer discussion thread.

docfail committed:
+ Moved custom replacement function from importFromByaf into util.js as sanitizeSafeCharacterReplacements() so that the replacement logic may be reused anywhere for consistency.

docfail committed:
> Missed a couple. Updated for consistency

Cohee1207 commented on src/endpoints/characters.js:
```diff
@@ -814,8 +814,69 @@ async function importFromByaf(uploadPath, { request }, preservedFileName) {
+            // BYAF does not support character expressions, so using the same structure will not result in conflicts,
+            // even if the expression system did not tolerate additional icons that are not mapped to expressions.
+            // This will not yet allow changing icons within the UI but at least the icons will be available for manual selection, rather than being lost.
+            const altImagesFolder = path.join(request.user.directories.characters, path.basename(fileName));
```
Sprites endpoint uses a sanitized card name instead of the file name to find images.
Here's the code for reference: https://github.com/SillyTavern/SillyTavern/blob/release/src/endpoints/sprites.js#L11
```suggestion
            const altImagesFolder = path.join(request.user.directories.characters, sanitize(card,name));
```

docfail replied:
shouldn't this be card.name?

docfail replied:
I'm fine with this, but for the record I don't love the idea of coupling functionality for finding existing files that we knew the name of at the time of saving to the use of the "sanitize" function, which isn't controlled by ST.

docfail replied:
Update, I have made this change. But without the ,

docfail replied:
Thanks for letting me know.

Cohee1207 replied:
Yeah my bad with the typo. You got it right.

docfail committed:
+ Preserve character full/display name without it needing to be safe as a filename.
> Change logic for naming the folder the alt images live in to match the logic the sprites endpoint uses to get character sprites for consistency

docfail committed:
> Only add full_name extension if displayName exists.

Cohee1207 committed:
full_name => display_name

docfail committed:
- No longer use safe lookalike replacements for filename sanitization to be consistent with how web browsers sanitize invalid characters in filenames before saving to disk.

Cohee1207 committed:
Tiny formatting fix

Cohee1207 commented:
We did it. I'll merge it now.

docfail commented:
Very happy to have this merged in. Thanks for the productive discussion.
